base/base/LineReader.cpp

```diff
@@ -36,33 +36,67 @@ bool hasInputData()
 
 }
 
-std::optional<LineReader::Suggest::WordsRange> LineReader::Suggest::getCompletions(const String & prefix, size_t prefix_length) const
+LineReader::Suggest::Words LineReader::Suggest::getCompletions(const String & prefix, size_t prefix_length)
 {
-    if (!ready)
-        return std::nullopt;
-
     std::string_view last_word;
 
     auto last_word_pos = prefix.find_last_of(word_break_characters);
     if (std::string::npos == last_word_pos)
         last_word = prefix;
     else
         last_word = std::string_view(prefix).substr(last_word_pos + 1, std::string::npos);
-
     /// last_word can be empty.
 
+    std::pair<Words::const_iterator, Words::const_iterator> range;
+
+    std::lock_guard lock(mutex);
+
     /// Only perform case sensitive completion when the prefix string contains any uppercase characters
     if (std::none_of(prefix.begin(), prefix.end(), [&](auto c) { return c >= 'A' && c <= 'Z'; }))
-        return std::equal_range(
+        range = std::equal_range(
             words_no_case.begin(), words_no_case.end(), last_word, [prefix_length](std::string_view s, std::string_view prefix_searched)
             {
                 return strncasecmp(s.data(), prefix_searched.data(), prefix_length) < 0;
             });
     else
-        return std::equal_range(words.begin(), words.end(), last_word, [prefix_length](std::string_view s, std::string_view prefix_searched)
+        range = std::equal_range(words.begin(), words.end(), last_word, [prefix_length](std::string_view s, std::string_view prefix_searched)
         {
             return strncmp(s.data(), prefix_searched.data(), prefix_length) < 0;
         });
+
+    return Words(range.first, range.second);
+}
+
+void LineReader::Suggest::addWords(Words && new_words)
+{
+    std::lock_guard lock(mutex);
+
+    /// Optimization for initial load.
+    if (words.empty())
+    {
+        words.swap(new_words);
+        words_no_case = words;
+    }
+    else
+    {
+        for (const auto & new_word : new_words)
+        {
+            if (std::binary_search(words.begin(), words.end(), new_word))
+                continue;
+
+            words.push_back(new_word);
+            words_no_case.push_back(new_word);
+        }
+    }
+
+    std::sort(words.begin(), words.end());
+    std::sort(words_no_case.begin(), words_no_case.end(), [](const std::string & str1, const std::string & str2)
+    {
+        return std::lexicographical_compare(begin(str1), end(str1), begin(str2), end(str2), [](const char char1, const char char2)
+        {
+            return std::tolower(char1) < std::tolower(char2);
+        });
+    });
```

Comment:
This way you're going to sort all the words again even if you've just added a single word to the list. We could instead put new words to the end of `words` without sorting, then sort only new words, after that use `std::inplace_merge` to get the full list sorted and finally call `std::unique()` to remove duplicates.

Reply:
I thought about `std::inplace_merge` initially, but decided that it does not worth it, since CREATE queries are not common.

But I like that after `std::inplace_merge`, `addWords()` can be used everywhere, so let's go this way.